### PR TITLE
fix(swingset): supervisor-xs: tolerate console.log(BigInt)

### LINKED ISSUE #2936: xs-worker doesn't tolerate BigInts in console.log()
## Describe the bug

In trying to get the XS worker usable for all vats, one problem I ran into was a vat that did `console.log()` of a BigInt. The xsnap supervisor is currently sending the log arguments to the supervisor with the same JSON.stringify it uses for all other messages (namely syscalls). This fails (somewhat opaquely) when vat code logs something that can't be serialized this way.

For the local worker, the vat's `console.log` is routed to two places:

https://github.com/Agoric/agoric-sdk/blob/fa68a8af6864d04a73fbc9dc70f63fb3d4225c1a/packages/SwingSet/src/kernel/kernel.js#L146-L155

The first is `makeConsole()`, which returns an instance of `anylogger`. The second is the `kernelSlot.vatConsole`, which wraps the log arguments and ultimately puts them into a serialized JSON object written to the slog file. That serialization special-cases any BigInts:

https://github.com/Agoric/agoric-sdk/blob/fa68a8af6864d04a73fbc9dc70f63fb3d4225c1a/packages/SwingSet/src/controller.js#L124-L136

For the XS worker, we need to handle this somehow. It's probably sufficient, for now, to just special-case BigInt (i.e. not use a full marshal, and accept the loss of fidelity for debug logs).

That probably means changing this code:

https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/src/kernel/vatManager/supervisor-subprocess-xsnap.js#L166-L168

to pre-transform `args`, like:

```js
port.send(['console', level, tag, ...(JSON.parse(JSON.stringify(args, (_, val) => { if (typeof(val) === 'bigint') { return Number(val); } return val; })))]);
```

In the long run, we need to nail down what exactly we're willing to tolerate in console.log calls, what precisely is the set of values that might appear in deliveries and syscalls, and make sure all our manager-worker communication pathways can support that full set. Our deliveries and syscalls are fairly conservative, and were designed to be JSON serializable, so it's really `console.log` that's the outlier.

## Testing

We should add a `console.log(0n)` to our test vat (in `test/workers/vat-target.js`) to exercise this.

### PR DESCRIPTION
**Pull request by dckc**
fixes #2936

### CODE REVIEW

File: packages/SwingSet/test/workers/vat-target.js [hunk 1]
```diff
@@ -15,6 +15,7 @@ function ignore(p) {
 
 export function buildRootObject(vatPowers, vatParameters) {
   console.log(`vat does buildRootObject`); // make sure console works
+  console.log(BigInt(0)); // make sure BigInt is tolerated: #2936
```
Comment (warner) [inline packages/SwingSet/test/workers/vat-target.js:18]: I just thought of a second test to add: `console.log` should not harden the arguments (which I've accidentally violated at least once). Please add something like:

```js
const mutable = [];
console.log('list:', mutable);
mutable.push(1);
console.log('list+1:', mutable);
```

to make sure I don't make that mistake again.

**Review by warner — APPROVED** [review]
Looks good to me. @erights how horrified are you to see the output of `q` parsed this way?

**Review by erights — APPROVED** [review]
LGTM.

In answer to @warner 's question:

> Looks good to me. @erights how horrified are you to see the output of `q` parsed this way?

Not at all horrified. What I should have remembered, and what this reminds me of, is that I purposely designed `q` to support this, after getting strong feedback from @FUDCo on the utility of JSON parseability even for data that code does not understand. The fact that this form is useful to you tells me a lot about the design constraints you're trying to meet.

I would still like to understand those constraints better. What is code *doing* with the parsed data despite the fact that the code does not understand the data? What is it doing such that a non-understood string was not adequate.

**Comment by dckc** [comment]
> ... What is code _doing_ with the parsed data despite the fact that the code does not understand the data? What is it doing such that a non-understood string was not adequate.

The code is providing a `console` to running vats. For most of the life of agoric-sdk, vats ran in the same node.js VM as the kernel, so the arguments that a vat gave to `console.log(...)` were passed as usual between JS functions up to the kernel, which passed them to node.js's `console.log` perhaps prepended with some sort of tag.

But when the vat runs in XS, the `console.log(...)` args from the vat don't get logged by the XS subprocess; they get sent up to the parent process. The naive implementation just `JSON.stringify`ed them. But then we started using bigints and the naive code went blammo.

**Comment by erights** [comment]
> > ... What is code _doing_ with the parsed data despite the fact that the code does not understand the data? What is it doing such that a non-understood string was not adequate.
> 
> The code is providing a `console` to running vats. For most of the life of agoric-sdk, vats ran in the same node.js VM as the kernel, so the arguments that a vat gave to `console.log(...)` were passed as usual between JS functions up to the kernel, which passed them to node.js's `console.log` perhaps prepended with some sort of tag.
> 
> But when the vat runs in XS, the `console.log(...)` args from the vat don't get logged by the XS subprocess; they get sent up to the parent process. The naive implementation just `JSON.stringify`ed them. But then we started using bigints and the naive code went blammo.

I understand why we needed to stop passing objects and pass only primitives instead. IOW, I understand the need for the `${q(stuff)}` I originally suggested. What I still do not understand is the value of `JSON.parse`ing it before giving it to the console. A simple string would equally well solve the problem you explain above.

**Comment by dckc** [comment]
> ... What I still do not understand is the value of `JSON.parse`ing it before giving it to the console.

I'm not entirely sure. I found @warner 's "would be nice" appealing, but when I thought about an argument behind it, I didn't come up with much:

> >  It would be nice to retain whatever structure we can, so I wouldn't want to entirely flatten/stringify the arguments.
 -- https://github.com/Agoric/agoric-sdk/issues/2936#issuecomment-827018210

**Pull request merged**